scripts/create_controller.py

```diff
@@ -0,0 +1,131 @@
+        """
+    )
+
+    controller_template = textwrap.dedent(
+        """\
+        /*
+         * This file is part of the NUbots Codebase.
+         *
+         * The NUbots Codebase is free software: you can redistribute it and/or modify
+         * it under the terms of the GNU General Public License as published by
+         * the Free Software Foundation, either version 3 of the License, or
+         * (at your option) any later version.
+         *
+         * The NUbots Codebase is distributed in the hope that it will be useful,
+         * but WITHOUT ANY WARRANTY; without even the implied warranty of
+         * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+         * GNU General Public License for more details.
+         *
+         * You should have received a copy of the GNU General Public License
+         * along with the NUbots Codebase.  If not, see <http://www.gnu.org/licenses/>.
+         *
+         * Copyright 2021 NUbots <nubots@newcastle.edu.au>
+         */
+
+        // You may need to add webots include files such as
+        // <webots/DistanceSensor.hpp>, <webots/Motor.hpp>, etc.
+        // and/or to add some other includes
+        #include <memory>
+        #include <webots/Robot.hpp>
+
+        #include "utility/tcp.hpp"
+
+        using namespace utility::tcp;
+
+        // This is the main program of your controller.
+        // It creates an instance of your Robot instance, launches its
+        // function(s) and destroys it at the end of the execution.
+        // Note that only one instance of Robot should be created in
+        // a controller program.
+        // The arguments of the main function can be specified by the
+        // "controllerArgs" field of the Robot node
+        int main(int argc, char** argv) {
+            // Make sure we have the command line arguments we need
+            if (argc != 3) {
+                std::cerr << "Usage: " << argv[0] << " <TCP PORT> <CONTROLLER_TIME_STEP>" << std::endl;
+                return 0;
+            }
+
+            // Load in the TCP port number from the command line and convert to an int
+            const int server_port = std::stoi(argv[1]);
+
+            // Load in the simulation timestep from the command line and convert to an int
+            const int time_step = std::stoi(argv[2]);
+
+            // Start the TCP server
+            int tcp_fd = create_socket_server(server_port);
+            if (tcp_fd == -1) {
+                std::cerr << "Error: Failed to start TCP server" << std::endl;
+                return 0;
+            }
+
+            // Create the Robot instance
+            std::unique_ptr<webots::Robot> robot = std::make_unique<webots::Robot>();
+
+            // Run the robot controller
+            while (robot->step(time_step) != -1) {
+                // Don't bother with doing anything unless we have an active TCP connection
+                if (tcp_fd == -1) {
```

Comment:
The value is changed inside the loop ... literally one line lower down

Reply:
well yes, but that's not my point. My point is that above all this is a check to see if `tcp_fd == -1`. If it is, then the program exits. So by the time we get to this loop, we know it's not. So this check is guaranteed to fail every single time it's checked, because to get this far, `tcp_fd != -1`, and besides the statements after this check is failed, `tcp_fd` won't change its value.
Basically the check before this loop, and the fact that besides this, `tcp_fd` isn't changed inside the loop mean that this check will never be true unless someone adds in code to the loop that changes `tcp_fd`.